In [10]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from IPython.display import display
import warnings
from sqlalchemy import create_engine
import pymysql
from matplotlib.patches import Patch

import ast

warnings.filterwarnings('ignore')

# 한글 폰트 설정
plt.rcParams['font.family'] = 'Malgun Gothic'

# 마이너스 기호 깨짐 방지
plt.rcParams['axes.unicode_minus'] = False

# 전역 시드 설정 (재현성을 위해)
np.random.seed(42)

print("="*60)
print("라이브러리 로드 완료!")
print("한글 폰트 설정 완료!")
print("="*60)

라이브러리 로드 완료!
한글 폰트 설정 완료!


In [11]:
engine = create_engine('mysql+pymysql://root:1215@localhost/review_analysis?charset=utf8mb4')

# cleaned_products 불러오기
cp = pd.read_sql("SELECT * FROM cleaned_products", engine)

# cleaned_reviews 불러오기
cr = pd.read_sql("SELECT * FROM cleaned_reviews", engine)

# df_v1 불러오기
df_v1 = pd.read_sql("SELECT * FROM df_v1", engine)

print("cleaned_products:", cp.shape)
print("cleaned_reviews:", cr.shape)
print("df_v1:", df_v1.shape)

cleaned_products: (2948, 17)
cleaned_reviews: (685292, 45)
df_v1: (685292, 61)


In [12]:
df = cr.copy()

In [13]:
# 리뷰 본문 NaN 제거
before = len(df)
df = df.dropna(subset=[TEXT_COL]).reset_index(drop=True)
print(f"NaN 제거: {before:,} → {len(df):,}건 (-{before-len(df):,})")

NaN 제거: 685,292 → 685,042건 (-250)


In [14]:
# 노이즈 확인하기
sample = df[TEXT_COL].astype(str).astype('object')

print("[노이즈 패턴 분포]")
print(f"엑셀 오류값(#NAME? 등): {sample.str.contains(r'#(NAME|N/A|VALUE|DIV/0|REF|NULL|NUM)', regex=True, na=False).sum():,}건")
print(f"HTML 태그(<br> 등): {sample.str.contains(r'<[^>]+>', regex=True, na=False).sum():,}건")
print(f"HTML 엔티티(&nbsp;): {sample.str.contains(r'&[a-z]+;', regex=True, na=False).sum():,}건")
print(f"URL: {sample.str.contains(r'https?://|www\.', regex=True, na=False).sum():,}건")
print(f"줄바꿈/탭: {sample.str.contains(r'[\r\n\t]', regex=True, na=False).sum():,}건")
print(f"반복문자 4회+: {sample.str.contains(r'(.)\1{3,}', regex=True, na=False).sum():,}건")
print(f"한글 없음: {(~sample.str.contains(r'[가-힣]', regex=True, na=False)).sum():,}건")

no_korean = df[~sample.str.contains(r'[가-힣]', regex=True, na=False)]
print(f"\n[한글 없는 리뷰 샘플]")
display(no_korean[TEXT_COL].head(10))

[노이즈 패턴 분포]
엑셀 오류값(#NAME? 등): 0건
HTML 태그(<br> 등): 234건
HTML 엔티티(&nbsp;): 2,167건
URL: 5건
줄바꿈/탭: 0건
반복문자 4회+: 5,994건
한글 없음: 184건

[한글 없는 리뷰 샘플]


20068                 oh holy shiiiit this best cargo bazi
34415                     very good very good gameverygood
36933           I love this one. I want to buy another one
44687                         Good quality and easy design
48243    the product's quality is pretty good and size ...
54051                     302 1159 6318 11302 1159 6318 11
54469                              12315152515261625616125
59645                                 14163662626277262671
65167    It is very good You have to buy this because o...
83771                                 ....................
Name: 리뷰내용, dtype: str

- 한글 없음 : 184건 모두 한국어 BERTopic 분석에서 어차피 제외되어야 함. 셀 4의 is_valid_for_topic 필터에서 자동으로 걸러짐 (한글 5자 이상)

In [15]:
from tqdm.auto import tqdm

# 통합 정제 함수
def clean_review_text(text):
    """
    무신사 리뷰 텍스트 1단계 전처리.
    임베딩(ko-sroberta) 입력용 정제 텍스트를 반환한다.
    """
    text = str(text)

    # (1) 엑셀 오류값 제거 (-> 0건)
    text = re.sub(r'#(NAME|N/A|VALUE|DIV/0|REF|NULL|NUM)[!?]?', ' ', text)

    # (2) URL 제거 -> (5건)
    text = re.sub(r'http\S+|www\.\S+', ' ', text)

    # (3) HTML 태그 + 엔티티 제거 -> (234건 + 2,167건)
    text = re.sub(r'<[^>]+>', ' ', text)
    text = re.sub(r'&[a-zA-Z]+;', ' ', text)

    # (4) 줄바꿈·탭 → 공백 -> (0건)
    text = re.sub(r'[\r\n\t]+', ' ', text)

    # (5) 반복문자 정규화
    text = re.sub(r'([ㄱ-ㅎㅏ-ㅣ])\1{2,}', r'\1\1', text) #같은 자음모음이 총 3번 이상 나오면 2번으로 줄임
    text = re.sub(r'(.)\1{3,}', r'\1\1', text) #한글(완성형), 영어, 숫자, 기호 등 아무 문자가 총 3번 이상 나오면 2번으로 줄임

    # (6) 연속 공백 정리
    text = re.sub(r'\s+', ' ', text).strip()

    return text

tqdm.pandas()
df['리뷰내용_clean'] = df[TEXT_COL].progress_apply(clean_review_text)
print("정제 완료")

  0%|          | 0/685042 [00:00<?, ?it/s]

정제 완료


- (5) 반복문자 정규화 한계 : 패턴반복(ex.와우와우와우)은 처리 안함. 추후 확인 후에 진행.

In [16]:
# 유효리뷰 플래그
df['한글_글자수'] = df['리뷰내용_clean'].str.count(r'[가-힣]')
df['전체_글자수'] = df['리뷰내용_clean'].str.len()

df['is_valid_for_topic'] = df['한글_글자수'] >= 5

print(f"[유효 리뷰 분포]")
print(f"  유효 (한글 5자 이상): {df['is_valid_for_topic'].sum():,}건 "
      f"({df['is_valid_for_topic'].mean()*100:.2f}%)")
print(f"  무효 (한글 5자 미만): {(~df['is_valid_for_topic']).sum():,}건")

print("\n[무효로 분류된 리뷰 샘플]")
display(df.loc[~df['is_valid_for_topic'], ['리뷰내용', '리뷰내용_clean', '한글_글자수']].head(10))

[유효 리뷰 분포]
  유효 (한글 5자 이상): 684,745건 (99.96%)
  무효 (한글 5자 미만): 297건

[무효로 분류된 리뷰 샘플]


,리뷰내용,리뷰내용_clean,한글_글자수
10387,굳,굳,1
20068,oh holy shiiiit this best cargo bazi,oh holy shiit this best cargo bazi,0
34415,very good very good gameverygood,very good very good gameverygood,0
36933,I love this one. I want to buy another one,I love this one. I want to buy another one,0
40978,이이이이이이이이이이이이이이이이이이이이이이이이이쁨,이이쁨,3
44174,굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿굿긋굿,굿굿긋굿,4
44687,Good quality and easy design,Good quality and easy design,0
48243,the product's quality is pretty good and size ...,the product's quality is pretty good and size ...,0
54051,302 1159 6318 11302 1159 6318 11,302 1159 6318 11302 1159 6318 11,0
54469,12315152515261625616125,12315152515261625616125,0


In [17]:
# 전후 비교검증
print("[정제 전후 비교 샘플]")
sample_idx = df[df['리뷰내용'] != df['리뷰내용_clean']].sample(10, random_state=42).index
for i in sample_idx:
    print(f"\n리뷰번호: {df.loc[i, '리뷰번호']}")
    print(f"  Before: {repr(df.loc[i, '리뷰내용'])[:120]}")
    print(f"  After : {repr(df.loc[i, '리뷰내용_clean'])[:120]}")

print("\n[글자 수 통계]")
print(df[['전체_글자수', '한글_글자수']].describe())

changed = (df['리뷰내용'].astype(str) != df['리뷰내용_clean']).sum()
print(f"\n정제로 인해 변경된 리뷰: {changed:,}건 ({changed/len(df)*100:.1f}%)")

[정제 전후 비교 샘플]

리뷰번호: 42995896
  Before: '깔끔하게 이너로 입기 좋아요 겨울이 오길  기다려짐'
  After : '깔끔하게 이너로 입기 좋아요 겨울이 오길 기다려짐'

리뷰번호: 4151073
  Before: '말그대로 롱티셔츠 입니다  웬만한 오버핏셔츠보다 살짝더 내려옵니다 과도하게 길게내려오진않고 2~4센치 내려오는 정도입니다  옆이 트여있어서 이뻐요 검은색과 반팔도 주문해달라네요 보드랍대요 재질이 너무 마음에 든답니
  After : '말그대로 롱티셔츠 입니다 웬만한 오버핏셔츠보다 살짝더 내려옵니다 과도하게 길게내려오진않고 2~4센치 내려오는 정도입니다 옆이 트여있어서 이뻐요 검은색과 반팔도 주문해달라네요 보드랍대요 재질이 너무 마음에 든답니다'

리뷰번호: 20155793
  Before: '생각보다 길고 이쁘네요  아무때나 걸쳐 입기 좋은거 같습니당.'
  After : '생각보다 길고 이쁘네요 아무때나 걸쳐 입기 좋은거 같습니당.'

리뷰번호: 52219604
  Before: '색감도 이쁘고 옷 입었을때 정핏으로 나와요. 오히려 좋아. '
  After : '색감도 이쁘고 옷 입었을때 정핏으로 나와요. 오히려 좋아.'

리뷰번호: 5750101
  Before: '이너용으로 입기 좋습니다 길이도 적당하고 두깨감도 적당합니다  가성비 참 좋은거 같습니다'
  After : '이너용으로 입기 좋습니다 길이도 적당하고 두깨감도 적당합니다 가성비 참 좋은거 같습니다'

리뷰번호: 6538251
  Before: '1. 원단 및 핏 원단 두께가 가을 ~ 봄 입을수 있을 듯 합니다. 핏은 부담스럽지 않은 와이드함 입니다. 제가 원하던 핏입니다.  2. 사이즈 180 / 77 34구입 제멋에서 구입한 와이드카고는 32 구입했는데
  After : '1. 원단 및 핏 원단 두께가 가을 ~ 봄 입을수 있을 듯 합니다. 핏은 부담스럽지 않은 와이드함 입니다. 제가 원하던 핏입니다. 2. 사이즈 180 / 77 34구입 제멋에서 구입

- min = 0(완전히 빈 리뷰)을 확인 할 수 있음. 해당 리뷰 확인필요

- 리뷰 내용 원본이 url 코드로 되어있음.
- 정제 후 : 빈 문자열
- 한글 글자수 = 0 이라 is_valid_for_topic = False로 자동 분류됨.
- 셀 2의 URL 포함 리뷰는 5건 -> 나머지 4건은 URL + 본문 텍스트로 예상됨.

In [18]:
# '글자 수 통계'의 min이 0인값 확인
empty_after = df[df['전체_글자수'] == 0]
print(f"정제 후 빈 리뷰: {len(empty_after):,}건")
print("\n[원본 확인]")
display(empty_after[['리뷰번호', '리뷰내용']].head(20))

정제 후 빈 리뷰: 1건

[원본 확인]


,리뷰번호,리뷰내용
236619,24482388,https://youtu.be/8ebPs7hKrc4


In [19]:
df_topic = df[df['is_valid_for_topic']].reset_index(drop=True)
docs = df_topic['리뷰내용_clean'].tolist()

print(f"BERTopic 입력 문서 수: {len(docs):,}건")
print(f"전체 대비 유효 비율: {len(docs)/len(df)*100:.2f}%")

output_path = f"./reviews_step1_cleaned.csv"
df.to_csv(output_path, index=False)
print(f"\n저장 완료: {output_path}")

BERTopic 입력 문서 수: 684,745건
전체 대비 유효 비율: 99.96%

저장 완료: ./reviews_step1_cleaned.csv
